### PR TITLE
enable search in Browse & Recording views

### LINKED ISSUE #11012: Use search bar in computer folders
Currently, the search bar is not available when browsing a folder in the "Computer" section of the library. (Tested with Mixxx 2.3.3 on Windows.)
This is problably due to the fact that the search bar only searches tracks which are added to the library database. As Mixxx reads and displays all metadata of the files in this view anyway as a sortable list, it should be no big issue to also search within that temporary list.

### PR DESCRIPTION
**Pull request by ronso0**
This enables a ~~simple~~ case-insensitive text filter for the Browse & Recording view via their proxymodels.
**Edit** not so simple anymore since with the last commit the query is not treated as one string but is split into words, so it allows queries like `holy "blue cow"`.

All columns are considered, incl. invisible.
Query is saved per feature only (not for individual folders).

closes #11012

### CODE REVIEW

**Comment by ronso0** [comment]
The last commit allows searching for individual words or quoted word groups.
Feels a bit like reinveting the wheel (sql column query) so this basic implementation is all I will do here.

File: src/library/proxytrackmodel.cpp [hunk 1]
```diff
@@ -147,26 +155,40 @@ bool ProxyTrackModel::filterAcceptsRow(int sourceRow,
         return false;
     }
 
+    if (m_currentSearch.trimmed().isEmpty()) {
+        return true;
+    }
+
+    QStringList tokens = SearchQueryParser::splitQueryIntoWords(
+            m_currentSearch.trimmed());
```
Comment (daschuer) [inline src/library/proxytrackmodel.cpp:163]: We already trim above. Please re-use the result.

File: src/library/proxytrackmodel.cpp [hunk 2]
```diff
@@ -147,26 +155,40 @@ bool ProxyTrackModel::filterAcceptsRow(int sourceRow,
...
-    while (!rowMatches && iter.hasNext()) {
-        int i = iter.next();
-        QModelIndex index = itemModel->index(sourceRow, i, sourceParent);
-        QVariant data = itemModel->data(index);
-        if (data.canConvert<QString>()) {
-            QString strData = data.toString();
-            if (strData.contains(filter)) {
-                rowMatches = true;
+
+    for (QString token : tokens) {
```
Comment (daschuer) [inline src/library/proxytrackmodel.cpp:170]:
```suggestion
    for (const auto& token : std::as_const(tokens)) {
```
Not sure if std::as_const() is strictly required though...

File: src/library/proxytrackmodel.cpp [hunk 3]
```diff
@@ -147,26 +155,40 @@ bool ProxyTrackModel::filterAcceptsRow(int sourceRow,
...
+        QListIterator<int> iter(filterColumns);
+        bool tokenMatch = false;
+        while (iter.hasNext() && !tokenMatch) {
```
Comment (daschuer) [inline src/library/proxytrackmodel.cpp:173]: This can also become a range based loop with a break statement below for early abbort.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for this first version. I have not yet tested it only some C++ suggestions.

How does the tokenizing compare to our normal search? I think it fails for things like "gimme gimme gimme" ... probably expected

**Comment by ronso0** [comment]
> How does the tokenizing compare to our normal search? I think it fails for things like "gimme gimme gimme" ... probably expected

Sorry, I don't understand the question, it's the same tokenizer used by `WTrackTableView::onSearch` > `SearchQueryParser::queryIsLessSpecific`.
And I don't know how to compare the basic text filter here and the sql filter in [`BaseTrackCache::filterAndSort`](https://github.com/mixxxdj/mixxx/blob/963f2ce9fce08ea10d40cbbbd1a8ac82a82d0a52/src/library/basetrackcache.cpp#L457) performance-wise if that's what you mean.

**Comment by daschuer** [comment]
Never mind, I have just tested it and it works like a charm. Thank you very much.

**Review by daschuer — APPROVED** [review]

**Pull request merged**